File: 프로그래머스/3/43164. 여행경로/여행경로.py
def solution(tickets):
    answer = []
    visited = [0]*len(tickets)
    
    def dfs(curr, res):
        if len(res) == len(tickets)+1:
            answer.append(res)
            return
        for idx in range(len(tickets)):
            if tickets[idx][0] == curr and visited[idx] == 0:
                visited[idx] = 1
                dfs(tickets[idx][1], res + [tickets[idx][1]])
                visited[idx] = 0
            
            
    dfs('ICN', ['ICN'])
    answer.sort()
    return answer[0]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    static String[][] tickets;
    static List<String> answer = new ArrayList<>();
    static int n;
    static boolean[] visited;
    
    public String[] solution(String[][] stickets) {
        tickets = stickets;
        n = tickets.length;
        visited = new boolean[n];
        
        dfs(0, "ICN", "ICN");
        Collections.sort(answer);
        
        return answer.get(0).split(",");
    }
    public static void dfs(int depth, String start, String path) {
            if (depth == n) {
                answer.add(path);
                return;
            }
            for (int i = 0; i < tickets.length; i++) {
                if (tickets[i][0].equals(start) && !visited[i]) {
                    visited[i] = true;
                    dfs(depth+1, tickets[i][1], path + ","+tickets[i][1]);
                    visited[i] = false;
                    
                }
            }
        }
}